### PR TITLE
network: refactor generic socket code

### PR DESCRIPTION
**Pull request by florincoras**
Move generic socket code into separate files. Refactor that prepares
move of socket apis from addresses to the socket class.

Risk Level: Low
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: include/envoy/network/socket.h [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+  SocketOptionName() = default;
+  SocketOptionName(const SocketOptionName&) = default;
+  SocketOptionName(int level, int option, const std::string& name)
+      : value_(std::make_tuple(level, option, name)) {}
+
+  int level() const { return std::get<0>(value_.value()); }
+  int option() const { return std::get<1>(value_.value()); }
+  const std::string& name() const { return std::get<2>(value_.value()); }
+
+  bool has_value() const { return value_.has_value(); }
```
Comment (jmarantz) [inline include/envoy/network/socket.h:33]: sorry, clang-tidy doesn't like this due to Envoy naming convention. This was in the codebase already but clang-tidy offers new levels of scrutiny when code moves to a new file.

You should probably just call this hasValue(), and update call-sites.

Comment (mattklein123) [reply include/envoy/network/socket.h:33]: +1 if you could pay it forward we would appreciate it.

Comment (florincoras) [reply include/envoy/network/socket.h]: No worries! Updated patch!

**Comment by mattklein123** [comment]
@florincoras please never force push as it makes reviews much more difficult. Please only merge master and add commits. Thank you!

**Comment by florincoras** [comment]
@mattklein123 apologies about that! Unfortunately, my lack of understanding of the PR process finally shows. I come from gerrit world where most of my development is done on branches that I keep on updating with new patch-sets. Will try to follow your suggestion. Thanks!

**Pull request merged**